components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/IdentityKeyStoreResolver.java

```diff
@@ -0,0 +1,374 @@
+import org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverConstants;
+import org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverConstants.InboundProtocol;
+import org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverException;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.security.interfaces.RSAPublicKey;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverUtil.buildCustomKeyStoreName;
+import static org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverUtil.buildTenantKeyStoreName;
+import static org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverUtil.getQNameWithIdentityNameSpace;
+
+
+/**
+ * KeyStore manager for inbound authentication protocols.
+ * Retrieve keystores, private keys, public keys and public certificates.
+ */
+public class IdentityKeyStoreResolver {
+
+    private static IdentityKeyStoreResolver instance = null;
+
+    private static ConcurrentHashMap<InboundProtocol, IdentityKeyStoreMapping>
+            keyStoreMappings = new ConcurrentHashMap<>();
+
+    // Hashmaps to store retrieved private keys and certificates.
+    // This will reduce the time required to read configs and load data from keystores everytime.
+    private static Map<String, Key> privateKeys = new ConcurrentHashMap<>();
+    private static Map<String, Certificate> publicCerts = new ConcurrentHashMap<>();
+
+    private static Log log = LogFactory.getLog(IdentityKeyStoreResolver.class);
+
+    private IdentityKeyStoreResolver() {
+
+        parseIdentityKeyStoreMappingConfigs();
+    }
+
+    public static IdentityKeyStoreResolver getInstance() {
+
+        if (instance == null) {
+            instance = new IdentityKeyStoreResolver();
+        }
+        return instance;
+    }
+
+    /**
+     * Return Primary or tenant keystore according to given tenant domain.
+     *
+     * @param tenantDomain  Tenant domain
+     * @return Primary or tenant keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public KeyStore getKeyStore(String tenantDomain) throws IdentityKeyStoreResolverException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        try {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                // Get primary keystore from keyStoreManager
+                return keyStoreManager.getPrimaryKeyStore();
+            }
+
+            // Get tenant keystore from keyStoreManager
+            String tenantKeyStoreName = buildTenantKeyStoreName(tenantDomain);
+            return keyStoreManager.getKeyStore(tenantKeyStoreName);
```

Comment:
There might be a possibility return null value for tenantKeyStoreName. Hence It would be great if there is having a NULL check.

Reply:
Here, the tenant key store name is directly derived from the tenant domain.
With the addition of the null check for method parameters, `tenantKeyStoreName` cannot be `null`